### DISCUSSION TITLE
Apicurio Registry - AVRO ref not working

### DISCUSSION THREAD

**Discussion by nagarashish** [Apicurio/apicurio-registry#4135] [Q&A] [answered]
Hello,

I am using apicurio/apicurio-registry-mem: latest release for understanding functionality of apicurio-registry.  The AVRO schema which is referencing other schema is not uploading.  Following are the sample AVRO schema files, uploaded in registry in the order listed below:

**Customer:**

```
{
  "namespace": "com.pizza.model",
  "type": "record",
  "name": "Customer",
  "fields": [
    {
      "name": "customerId",
      "type": "string"
    },
    {
      "name": "name",
      "type": "string"
    },
    {
      "name": "address",
      "type": "string"
    },
    {
      "name": "phone",
      "type": "string"
    }
  ]
}
```
**Pizza Item**
```
{
  "namespace": "com.pizza.model",
  "type": "record",
  "name": "PizzaItem",
  "fields": [
    {
      "name": "pizzaId",
      "type": "string"
    },
    {
      "name": "type",
      "type": "string"
    },
    {
      "name": "size",
      "type": "string"
    },
    {
      "name": "toppings",
      "type": {
        "type": "array",
        "items": "string"
      }
    },
    {
      "name": "quantity",
      "type": "int"
    }
  ]
}
```
I am trying to upload following schema which references above 2 schemas:
```
{
  "namespace": "com.pizza.model",
  "type": "record",
  "name": "PizzaOrder",
  "fields": [
    {"name": "orderId", "type": "string"},
    {"name": "customer", "type": "com.pizza.model.Customer"},
    {"name": "items", "type": {"type": "array", "items": "com.pizza.model.PizzaItem"}},
    {"name": "totalPrice", "type": "double"},
    {"name": "paymentStatus", "type": "string"},
    {"name": "deliveryStatus", "type": "string"},
    {"name": "orderTime", "type": "long"}
  ]
}
```
I get following error:

**schemaParseException: "com.pizza.model.Customer" is not a defined name. The type of the "customer" field must be a defined name or a {"type": ...} expression.**

Not sure how to resolve this error.  I have made sure all the schemas which are referenced are already uploaded in the registry.
Is there any document or guideline to follow?  In the global rules, I have set validity rule to FULL and Integrity rule to "Refs must exist".  
 
Thanks for helping in advance.

**Comment by EricWittmann** [comment, marked as answer]
The Registry does support references from one Avro schema to other schemas, and you've got one part of it right (by making sure the `Customer` and `PizzaItem` schemas are already uploaded).  But the missing bit is that when you register the `PizzaOrder` schema you need to provide **mappings** to the references it includes.

What does this mean?  It means that when you register `PizzaOrder` you need to tell Registry where `com.pizza.model.PizzaItem` exists **and** where `com.pizza.model.Customer` exists.  You do this by providing mappings from those names to coordinates in the Registry (groupId, artifactId, version).

Documentation on this can be found here:

https://www.apicur.io/registry/docs/apicurio-registry/2.5.x/getting-started/assembly-managing-registry-artifacts-api.html#managing-artifact-references-using-rest-api_registry

**Reply by davidradl** [reply]
@EricWittmann do you mean  `It means that when you register PizzaOrder` not `It means that when you register PizzaItem`.
I assume by mappings you mean supplying references property in the parent json object in addition to the content ; `Item` in the case of the docs example, `PizzaOrder` in this example.  Have I understood this correctly?

**Reply by EricWittmann** [reply]
Yes you are right!  Fixed. :)

**Reply by EricWittmann** [reply]
And yes, you understand correctly I think.
